### PR TITLE
[On-chain, Relayminer] chore: add claim msg validation

### PR DESCRIPTION
**Pull request by bryanchriswhite**
### Human Summary

Adds validation logic around the claim's session during claim creation. The session ID must match that as determined by an on-chain query of the session given the same inputs. Additionally, the supplier list for the session must contain the supplier whose address is referenced in the claim.

### Summary generated by Reviewpad on 07 Dec 23 10:58 UTC
This pull request includes various changes across multiple files. Here is a summary of the changes:

- `supplier.go` (in `testutil/keeper` package):
  - Added imports for `supplier` and `supplier/mocks` packages.
  - Added import for `sessiontypes` package from `github.com/pokt-network/poktroll/x/session/types`.
  - Modified the `SupplierKeeper` function to accept an additional parameter `sessionByAppAddr` of type `supplier.SessionsByAppAddress`.
  - Added implementation for `GetSession` method of the `mockSessionKeeper` mock.
  - Added a call to `k.SupplySessionKeeper(mockSessionKeeper)`.

- `relay_verifier.go` (in `pkg/relayer/proxy` package):
  - Added import for `sdkerrors` package.
  - Added import for `ring_secp256k1` package.
  - Removed import for `github.com/cometbft/cometbft/crypto` package.
  - Removed import for `github.com/pokt-network/poktroll/x/session/types` package.
  - Modified the `VerifyRelayRequest` function to handle missing application address from the relay request.
  - Changed the method of obtaining signable bytes of the relay request to `GetSignableBytesHash`.
  - Changed the method of verifying relay request's signature using `ringSig`.
  - Modified the code for verifying relay request session using `GetSession` method of `rp.sessionQuerier`.
  - Improved error message when session ID does not match relay request's session ID.

- `relay_builders.go`:
  - Added a debug log message when there is a failure in unmarshaling the relay request.

- `ring_signer.go`:
  - Changed the `Sign` function to accept an array of 32 bytes for the message argument instead of a byte slice.

- `sessiontypes` package:
  - Added `SessionKeeper` interface which includes the `GetSession` method.
  - Updated the `go:generate mockgen` command to include the `SessionKeeper` interface in the mock generation.

- `testqueryclients/appquerier.go`:
  - Updated the import paths for the `mockclient` and `types` packages.
  - Replaced the context parameter in the `GetApplication` function with an underscore.
  - Updated the return type of the `GetApplication` function to use the `apptypes` type.
  - Updated the variable names within the `GetApplication` function.
  - Updated the error type returned in case of an application not found to `apptypes.ErrAppNotFound`.

- `app.go`:
  - Changes to the `New` function related to the ordering of code and addition of comments.

- `interface.go`:
  - Added `SupplierQueryClient` interface.
  - Added `SessionQueryClient` interface.

- `session.feature`:
  - Added a new feature file describing the "Session Namespace" feature.
  - Includes a scenario for completing the claim/proof lifecycle for a valid session.

- `simple_signer.go`:
  - Changed the signature parameter of the `Sign` function to accept an array of bytes with a fixed size of 32.

- `logger.go` (in `pkg/polylog/polyzero` package):
  - Added the `os` package as an import.
  - Updated the `NewLogger` function to provide default logger configuration.
  - Added TODO comments for improving and adding new logger functions.

- `client.go`:
  - Added import of the `cosmossdk.io/api/tendermint/abci` package.
  - Removed import of the `abciTypes` package from `github.com/cometbft/cometbft/abci/types`.
  - Changed the type `TxEvent` to be an alias for the `abci.TxResult` type.
  - Added comments.

- `proxy.go`:
  - Removed the import of `cosmosclient` package.
  - Removed the import of `querytypes` package.
  - Replaced the type `suppliertypes.QueryClient` with `client.SupplierQueryClient`.
  - Replaced the type `sessiontypes.QueryClient` with `client.SessionQueryClient`.
  - Removed the field `clientCtx`.
  - Replaced the initialization of `supplierQuerier` with `client.SupplierQueryClient`.
  - Replaced the initialization of `sessionQuerier` with `client.SessionQueryClient`.
  - Removed the initialization of `keyring`.
  - Added a check to ensure `proxiedServicesEndpoints` is not empty in the `validateConfig` method.

- `relayerproxy.go`:
  - Added various package imports and new struct and functions related to testing utilities and mock structures for the relayer proxy.

## Issue

- #140 
- #237 

On-chain claim message handling should validate that the claimed session matches the session at the given claimed sesison end height; i.e. given some session inputs, ensure that the ID matches and that the supplier is in the session.

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [ ] **Run all unit tests**: `make go_develop_and_test`
- [ ] **Verify Localnet manually**: See the instructions [here](TODO: add link to instructions)

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

File: testutil/keeper/supplier.go
Comment (bryanchriswhite) [inline testutil/keeper/supplier.go:75]: Where would these integers be better sourced from?

**Review by Olshansk — CHANGES_REQUESTED** [review]
Posting a partial review and will finish tomorrow.

File: e2e/tests/session.feature
Comment (bryanchriswhite) [reply e2e/tests/session.feature:6]: I think the "when" should line up with the action, which I think is actually the previous step. I also think it makes sense for the action to be in present tense. Then, with an "and" in front instead, "after..." reads much  better IMO. I feel like the "after" is a really clear way to communicate that this step will be waiting for something.

```gherkin
Scenario: Supplier completes claim/proof lifecycle for a valid session
    Given the user has the pocketd binary installed
    When the supplier "supplier1" services a session with "5" relays for service "anvil" for application "app1"
    And after the supplier creates a claim for the session
```

File: e2e/tests/session_steps_test.go
Comment (bryanchriswhite) [reply e2e/tests/session_steps_test.go:116]: Yes, #220

File: testutil/keeper/supplier.go
Comment (bryanchriswhite) [reply testutil/keeper/supplier.go:75]: Sorry, I'm not following. Can you elaborate?

File: e2e/tests/session_steps_test.go
Comment (bryanchriswhite) [inline e2e/tests/session_steps_test.go:110]:
```suggestion
	// TODO_IN_THIS_COMMIT: use testclient.CometLocalWebsocketURL instead.
	pocketNodeWebsocketUrl := "ws://poktroll-sequencer:36657/websocket"
```

File: x/session/keeper/session_hydrator.go
Comment (Olshansk) [inline x/session/keeper/session_hydrator.go:161]: Ty for catching htis!

File: x/supplier/client/cli/query_claim_test.go
Comment (Olshansk) [inline x/supplier/client/cli/query_claim_test.go:148]: `as we expect to need for the test`

Can `20` be a function of all the counts provided to the function?

File: x/supplier/client/cli/query_claim_test.go
Comment (Olshansk) [inline x/supplier/client/cli/query_claim_test.go:211]: Weird that the bot didn't pick this up

File: x/supplier/client/cli/query_claim_test.go
Comment (Olshansk) [inline x/supplier/client/cli/query_claim_test.go:212]: NIT: loop through apps on the outside.

Fundamentally it results in the same thing, but logically it makes more sense given the `1 app -> N suppliers` relationship in a session.

File: x/supplier/client/cli/query_claim_test.go
Comment (Olshansk) [inline x/supplier/client/cli/query_claim_test.go:221]: You have not idea how much time I wasted on this...

File: x/supplier/client/cli/query_claim_test.go
Comment (Olshansk) [inline x/supplier/client/cli/query_claim_test.go:235]: TODO_IN_THIS_PR?

File: x/supplier/keeper/msg_server_create_claim_test.go
Comment (Olshansk) [inline x/supplier/keeper/msg_server_create_claim_test.go:22]: Per one of my other comments. Reconsider: s/mock/fixture or the use of this structure at all.

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: e2e/tests/session.feature
Comment (bryanchriswhite) [reply e2e/tests/session.feature:4]: Would you mind if we do  this in a separate PR as it will require updating all existing feature files as well.

File: testutil/keeper/supplier.go
Comment (bryanchriswhite) [reply testutil/keeper/supplier.go]: I've refactored to not need a new type. It seems like that was mostly what you were getting at.

File: testutil/testclient/localnet.go
Comment (bryanchriswhite) [inline testutil/testclient/localnet.go:70]: @okdas (same as https://github.com/pokt-network/poktroll/pull/236#discussion_r1413781028)

File: pkg/client/tx/client.go
Comment (Olshansk) [inline pkg/client/tx/client.go:93]:
```suggestion
	// TxResult is the CommetBFT result of committing a transaction on-chain
	Result TxResult `json:"result"`
```

File: x/supplier/keeper/msg_server_create_claim.go
Comment (Olshansk) [inline x/supplier/keeper/msg_server_create_claim.go:63]: Just a heads up that it'll be a couple rounds (because others have PRs in flight as well)

File: x/supplier/keeper/msg_server_create_claim.go
Comment (Olshansk) [inline x/supplier/keeper/msg_server_create_claim.go:103]:
```suggestion
	// TODO: return the claim in the response.
```

File: x/supplier/client/cli/helpers_test.go
Comment (Olshansk) [inline x/supplier/client/cli/helpers_test.go:16]: NIT: Introduce semantically named types

File: x/supplier/client/cli/helpers_test.go
Comment (Olshansk) [inline x/supplier/client/cli/helpers_test.go:68]: TODO_IN_THIS_COMMIT. 

Still intending todo in this commit?

File: testutil/testclient/localnet.go
Comment (okdas) [reply testutil/testclient/localnet.go:19]: @bryanchriswhite yep! That hostname will now resolve to sequencer on both LocalNet and DevNet!

File: x/supplier/client/cli/helpers_test.go
Comment (Olshansk) [inline x/supplier/client/cli/helpers_test.go:162]:
```suggestion
// encodeSessionHeader returns a base64 encoded string of a json
// serialized session header.
func encodeSessionHeader(
```

File: x/supplier/client/cli/helpers_test.go
Comment (Olshansk) [inline x/supplier/client/cli/helpers_test.go:214]:
```suggestion
	// TODO_TECHDEBT:  Forward the actual claim in the response once the response is updated to return it.
	return &types.Claim{
```

File: x/supplier/client/cli/helpers_test.go
Comment (Olshansk) [inline x/supplier/client/cli/helpers_test.go:182]:
```suggestion
// createClaim sends a tx using the test CLI to create an on-chain claim
func createClaim(
```

File: testutil/testclient/localnet.go
Comment (Olshansk) [reply testutil/testclient/localnet.go:19]: @okdas Bump

File: pkg/client/tx/client.go
Comment (Olshansk) [inline pkg/client/tx/client.go:96]: I'm worried that `TxResult` may be conflated with the definition of a TxResult in Comet/Tendermint land. Wdyt if `ABCIEvents`, `TxResultEvents` or removing the decoupling altogether?

Screenshot 2023-12-04 at 3 07 34 PM

https://github.com/cosmos/cosmos-sdk/blob/main/proto/tendermint/abci/types.proto#L416

File: e2e/tests/session_steps_test.go
Comment (Olshansk) [inline e2e/tests/session_steps_test.go:101]: 👍

File: testutil/supplier/fixtures.go
Comment (Olshansk) [inline testutil/supplier/fixtures.go:32]: I think this warrants a helper struct. It'll make the business logic cleaner. 

If you want to get this over the finish line though, I'd be okay with a `TODO_TECHDEBT/TODO_COMMUNITY`

File: testutil/supplier/fixtures.go
Comment (Olshansk) [inline testutil/supplier/fixtures.go:78]:
```suggestion
// newSuppliers configures a supplier for the services provided and nil endpoints.
func newSupplier(t *testing.T, addr string, services ...*sharedtypes.Service) *sharedtypes.Supplier {
```

File: testutil/supplier/fixtures.go
Comment (Olshansk) [inline testutil/supplier/fixtures.go:96]:
```suggestion
// newApplication configures a supplier for the services provided.
func newApplication(t *testing.T, addr string, services ...*sharedtypes.Service) *apptypes.Application {
```

File: x/supplier/client/cli/helpers_test.go
Comment (Olshansk) [inline x/supplier/client/cli/helpers_test.go:222]:
```suggestion
// getSessionId sends a query using the test CLI to get a session for the inputs provided.
// It is assumed that the supplierAddr will be in that session based on the test design, but this
// is insured in this function before it's successfully returned.
func getSessionId(
```

File: x/supplier/client/cli/helpers_test.go
Comment (Olshansk) [inline x/supplier/client/cli/helpers_test.go:235]: Thoughts on having the caller pass in this local constant?

**Review by Olshansk — CHANGES_REQUESTED** [review]
The PR is looking really good overall. Really appreciate both the changes, and comments to set up followup work.

Left a few lingering comments but should be g2g right after!

File: pkg/client/tx/client.go
Comment (bryanchriswhite) [reply pkg/client/tx/client.go:93]: Tyty :pray:

File: x/supplier/client/cli/helpers_test.go
Comment (bryanchriswhite) [reply x/supplier/client/cli/helpers_test.go:16]: Nice catch, this was probably my IDE doing something dumb when I refactored some things.

File: x/supplier/client/cli/helpers_test.go
Comment (bryanchriswhite) [reply x/supplier/client/cli/helpers_test.go:68]: I'm gonna remove the comment and the unused `serviceCount` param. My original thinking was that it would make sense for this function to parameterize services as well, but the need hasn't quite arisen yet so I would put it off at this point.

File: x/supplier/client/cli/helpers_test.go
Comment (bryanchriswhite) [reply x/supplier/client/cli/helpers_test.go:235]: Definitely have thoughts :sweat_smile:. This is related to https://github.com/pokt-network/poktroll/pull/236#discussion_r1415261843. 

TL;DR, I would  hold off for now.

It might be better to take a step back and think about how we would like multi-service integration tests to be designed, and then evaluate if it's compatible with our current direction (i.e. refactor and/or reuse what we have) or if we need to do that in a new test suite with more comprehensive helpers. Wdyt?

File: pkg/client/tx/client.go
Comment (bryanchriswhite) [reply pkg/client/tx/client.go:97]: Thanks for calling this out! :raised_hands:  This is indeed the correct type and we should be using it instead of the one(s) currently defined in our tx client package. :mechanical_arm: 

I have gone digging (on more than one occasion) for the actual protobuf that is being used by comet/cosmos to produce the JSON structure we see in the event subscription stream. I can't remember if I looked at this one before or not but I'm realizing that it's tricky to connect the dots because it's being serialized to JSON. :thinking:  Anyways, you nailed it - thanks!

File: testutil/supplier/fixtures.go
Comment (bryanchriswhite) [reply testutil/supplier/fixtures.go:32]: Good call. I found a  low-verbosity solution. :+1:

File: pkg/client/tx/client.go
Comment (Olshansk) [inline pkg/client/tx/client.go:90]: So much cleaner!

File: testutil/supplier/fixtures.go
Comment (Olshansk) [inline testutil/supplier/fixtures.go:24]: Ty for making the change!

File: x/supplier/client/cli/helpers_test.go
Comment (Olshansk) [reply x/supplier/client/cli/helpers_test.go:240]: I think holding off and seeing what/when we need from multi-service integration makes sense.

Let's hold off 👍

**Review by Olshansk — APPROVED** [review]
LGTM. The last round of comments & cleanup really put the :bowtie: on it!

Note that the E2E tests are failing and wanted to make sure you remember the base branch here is not `main` (should it be?), but otherwise g2g!

**Comment by bryanchriswhite** [comment]
> Note that the E2E tests are failing and wanted to make sure you remember the base branch here is not `main` (should it be?), but otherwise g2g!

Thanks for calling this out! I typically intentionally keep the github base of PR branches the same as the git (actual) base until after approval to avoid cluttering the commit list during review. I always take note of the +/- LOC diff before and after re-github-basing (not an actual rebase operation). In the event of discrepancies and/or merge conflict resolutions I double-check everything before merging and would ask for re-review if something unexpected arises or changes since approval.

File: x/supplier/client/cli/helpers_test.go
Comment (bryanchriswhite) [reply x/supplier/client/cli/helpers_test.go:240]: :pray:

**Pull request merged**
